### ISSUE TITLE
`Instance` raises locality warning at `Qed` / `Defined` time.

### ISSUE THREAD

**Issue by Zimmi48** [coq/coq#15030]
#### Description of the problem

When defining instances in proof mode (example below), the warning for the missing locality attribute is triggered at `Qed` / `Defined` time instead of being triggered at instance-definition time:

```coq
Class foo := { number : nat }.

Instance bar : foo. (* This is where the warning is triggered in 8.15. *)
split.
exact 0.
Qed. (* This is where the warning is triggered in 8.14. *)
```

This is problematic if someone wants to follow an automatic porting approach like the one recommended at https://coq.discourse.group/t/change-of-default-locality-for-hint-commands-in-coq-8-13/1140/13

#### Coq Version

8.14.0

The problem is already solved in 8.15 / master, so it might just be a matter of backporting the relevant PR. If it's too hard to backport, I'd recommend removing the deprecation warning in 8.14.1 so that users only have to face it when it's become easy to fix automatically...

cc @silene

**Comment by Zimmi48** [comment]
Thanks for pointing out that the problem was already reported. This report is about raising the question that the warning without the fix may be problematic to keep in 8.14.

**Comment by Zimmi48** [comment]
Are you talking about the "quick & dirty" script from https://coq.discourse.group/t/change-of-default-locality-for-hint-commands-in-coq-8-13/1140? Then, just adapting it to `Instance` has an issue which is that it will also add `#[ global ]` on instances within sections, but contrary to the case of `Hint`, these won't be rejected by Coq (but they will still change the overall behavior).

**Comment by Zimmi48** [comment]
Oh that's great! Could you post a link to this script in the Discourse thread and update the top post to make it easy to find?

**Comment by ppedrot** [comment]
The script is [actually already available here](https://github.com/ppedrot/coq/blob/warn-unqualified-typeclass-instance-script/adapt.ml) but I am not sure I want to advertise it since it's quite hackish. It worked good enough to port the stdlib but it lacks any kind of infrastructure, i.e. you have to print the stderr output of your process somewhere and give that to the script in a way that the filepaths make sense.

**Comment by Zimmi48** [comment]
It's better than any of the currently advertised solutions for sure, so it deserves to be advertised at least at the same level as these.

**Comment by yforster** [comment]
I'm getting `Exception: Invalid_argument "String.sub / Bytes.sub".` when using the script. @ppedrot do you have more instructions how to run it? I printed `stderr` to a file `error`, did `chmod +x adapt.ml` and then` ./adapt.ml error` in a context where the paths in the error messages should work.

**Comment by ppedrot** [comment]
Hmm, there are some hardcoded lengths in the script, but I don't remember what they were for. Try tweaking that so that it extracts the file correctly, maybe using a regexp to be more robust.

**Comment by yforster** [comment]
Thanks, that was it! The `let f = String.sub f 20 (String.length f - 20) in` was removing something which didn't need to be removed in my setting, so simply commenting the line made the script work.

**Comment by Zimmi48** [comment]
So can the Discourse post be updated with info on how to properly handle `Instance` locality warnings?

**Comment by ppedrot** [comment]
Dunno. @yforster did the script give reasonable outputs?

**Comment by RalfJung** [comment]
I agree that either the line number fix should be backported to 8.14 or the warning disabled in 8.14. This warning is causing a lot of pain in all our codebases, and the fact that Coq does not ship with a way to automatically resolve the warning means everyone assembles their own hacky script with different quirks -- and most of them will fail due to the line number issue.

**Comment by RalfJung** [comment]
> So can the Discourse post be updated with info on how to properly handle Instance locality warnings?
>
> Dunno. @yforster did the script give reasonable outputs?

Looking at [the discourse thread](https://coq.discourse.group/t/change-of-default-locality-for-hint-commands-in-coq-8-13/1140) I do not see instructions for how to handle the `Instance` warnings that started with 8.14. There is Tej's script but that does not work due to the line number bug. @ppedrot  Am I looking at the wrong thread?

IMO a tool to fix the warning should be shipped with Coq itself, since this will affect basically all existing codebases.

**Comment by SkySkimmer** [comment]
The change in master is imo too big for backport, but we can do https://github.com/coq/coq/pull/15243

**Comment by RalfJung** [comment]
I don't understand what "even when interactive" has to do with incorrect line numbers, but if this means correct line numbers are printed then I am all in favor. :)

**Comment by Blaisorblade** [comment]
> I don't understand what "even when interactive" has to do with incorrect line numbers, but if this means correct line numbers are printed then I am all in favor. :)

It does mean that, according to
https://github.com/coq/coq/blob/13a9c21215d882323162a919c682b2fcda4a0b58/doc/changelog/07-vernac-commands-and-options/15243-instance-locality.rst.

Until that's merged:

> There is also a [link to an .ml file](https://github.com/ppedrot/coq/blob/warn-unqualified-typeclass-instance-script/adapt.ml) but I have no idea how to 'run' an .ml file as a script, nor is there any comment explaining how this file needs to be invoked.

You happen to know what `#!/usr/bin/env ocaml` means, no? https://github.com/coq/coq/issues/15030#issuecomment-964293819 and https://github.com/coq/coq/issues/15030#issuecomment-946984417 say the rest.

**Comment by RalfJung** [comment]
Thanks! When I try that script on the output in https://gitlab.mpi-sws.org/iris/simuliris/-/jobs/159456 I get
```
$ ~/src/coq/instance-hint-adapt.ml warn
File "/home/r/src/coq/instance-hint-adapt.ml", line 64, characters 24-42:
64 |   let todo = List.sort (Pervasives.compare : int -> int -> int) todo in
                             ^^^^^^^^^^^^^^^^^^
Alert deprecated: module Stdlib.Pervasives
Use Stdlib instead.

If you need to stay compatible with OCaml < 4.07, you can use the 
stdlib-shims library: https://github.com/ocaml/stdlib-shims
Exception: Sys_error "behavior.v: No such file or directory".
```
The file is called `theories/simulang/behavior.v` and the error output uses that path but somehow that does not seem to be picked up correctly?

Which assumptions does the script make about the warning output I feed it -- will it also take care of other similar deprecation warnings for `Hint` (not `Instance`)? What happens when there are other unrelated warnings in the output?

**Comment by ppedrot** [comment]
This is fixed both in 8.14 and master, so we can safely close.
